File: ros2_ws/src/flexiv_driver/flexiv_driver/arm_commander.py
"""ArmCommander — Single owner of the Flexiv RDK connection.

Pure Python, no ROS2 dependency. All RDK access is lock-protected so a
watchdog thread (Step 3) can safely share the connection.

Usage:
    from common.config import load_config
    from flexiv_driver.arm_commander import ArmCommander

    config = load_config("config/robot.yaml")
    arm = ArmCommander(config)
    arm.connect()
    arm.home()
    arm.zero_ft()
    arm.stream_cartesian(some_pose)
    arm.shutdown()
"""

from __future__ import annotations

import logging
import threading
import time
from typing import Optional

import flexivrdk

from common.types import Pose, RobotState, Wrench
from common.config import RobotConfig
from common.errors import ConnectionLost, RobotFault, SafetyViolation
from flexiv_driver.safety import SafetyChecker, SafetyWatchdog

logger = logging.getLogger(__name__)


class ArmCommander:
    """Single owner of the Flexiv RDK robot connection.

    All methods are thread-safe via an internal lock. Mode switching is
    handled internally — callers never need to know about RDK modes.
    """

    def __init__(self, config: RobotConfig):
        self._config = config
        self._robot: Optional[flexivrdk.Robot] = None
        self._tool: Optional[flexivrdk.Tool] = None
        self._lock = threading.Lock()
        self._mode: Optional[flexivrdk.Mode] = None
        self._connected = False
        self._stopped = False
        self._safety = SafetyChecker(config.safety)
        self._watchdog: Optional[SafetyWatchdog] = None
        self._configured_max_contact_wrench: Optional[tuple[float, ...]] = None

    # ──────────────────────────────────────────────────────────────────
    # Connection lifecycle
    # ──────────────────────────────────────────────────────────────────

    def connect(self):
        """Connect to robot, clear faults, enable, wait for operational."""
        with self._lock:
            self._robot = flexivrdk.Robot(self._config.serial_number)

            if self._robot.fault():
                logger.warning("Fault detected — clearing...")
                if not self._robot.ClearFault():
                    raise RobotFault("Cannot clear robot fault")
                time.sleep(2)
                if self._robot.fault():
                    raise RobotFault(
                        "Cannot clear robot fault — check E-stop and teach pendant"
                    )
                logger.info("Fault cleared")

            logger.info("Enabling robot...")
            self._robot.Enable()

            timeout = 10.0
            start = time.monotonic()
            while not self._robot.operational():
                if time.monotonic() - start > timeout:
                    raise ConnectionLost("Timeout waiting for robot to become operational")
                time.sleep(0.1)

            self._tool = flexivrdk.Tool(self._robot)
            self._mode = None
            self._connected = True
            self._stopped = False
            self._configured_max_contact_wrench = None
            logger.info("Robot operational: %s", self._config.serial_number)

        self._start_watchdog()

    def shutdown(self):
        """Stop robot and release connection."""
        self._stop_watchdog()
        with self._lock:
            if self._robot is not None:
                try:
                    self._robot.Stop()
                except Exception as e:
                    logger.warning("Error during shutdown Stop(): %s", e)
            self._connected = False
            self._stopped = True
            logger.info("Robot shutdown complete")

    # ──────────────────────────────────────────────────────────────────
    # Blocking primitives
    # ──────────────────────────────────────────────────────────────────

    def home(self):
        """Move to home position via built-in PLAN-Home. Blocking."""
        self._ensure_connected()
        self._pause_stream_heartbeat()
        with self._lock:
            self._switch_mode(flexivrdk.Mode.NRT_PLAN_EXECUTION)
            self._robot.ExecutePlan("PLAN-Home")

        # Poll outside the lock so watchdog can still read state
        while True:
            with self._lock:
                if not self._robot.busy():
                    break
            time.sleep(0.1)
        logger.info("Home position reached")

    def move_to(self, pose: Pose, velocity: float = 0.05):
        """Blocking MoveL to target pose.

        Args:
            pose: Target pose (ROS2 convention internally, converted to RDK).
            velocity: TCP linear velocity in m/s.
        """
        self._ensure_connected()
        self._check_pose_safe(pose)
        self._pause_stream_heartbeat()

        # MoveL expects Coord with [mm,mm,mm] position and [deg,deg,deg] rotation.
        # But actually RDK Coord takes metres and degrees depending on the primitive.
        # From scan_node: it passes metres for position, degrees for rotation.
        from scipy.spatial.transform import Rotation as R

        quat_xyzw = [pose.qx, pose.qy, pose.qz, pose.qw]
        euler_deg = R.from_quat(quat_xyzw).as_euler("xyz", degrees=True).tolist()
        pos_m = [pose.x, pose.y, pose.z]

        with self._lock:
            self._switch_mode(flexivrdk.Mode.NRT_PRIMITIVE_EXECUTION)
            self._robot.ExecutePrimitive(
                "MoveL",
                {
                    "target": flexivrdk.Coord(
                        pos_m, euler_deg, ["WORLD", "WORLD_ORIGIN"]
                    ),
                    "vel": velocity,
                },
            )

        # Poll until reached
        while True:
            with self._lock:
                states = self._robot.primitive_states()
            if states.get("reachedTarget"):
                break
            time.sleep(0.1)
        logger.info("MoveL complete")

    def zero_ft(self):
        """Zero the force/torque sensor. Blocking."""
        self._ensure_connected()
        self._pause_stream_heartbeat()
        with self._lock:
            self._switch_mode(flexivrdk.Mode.NRT_PRIMITIVE_EXECUTION)
            self._robot.ExecutePrimitive("ZeroFTSensor", dict())

        while True:
            with self._lock:
                states = self._robot.primitive_states()
            if states.get("terminated"):
                break
            time.sleep(0.1)
        logger.info("FT sensor zeroed")

    def contact(
        self,
        direction: list[float] = None,
        velocity: float = 0.02,
        max_force: float = 20.0,
    ):
        """Move in direction until contact. Blocking.

        Uses Flexiv's built-in Contact primitive — firmware-level,
        stops automatically when force exceeds max_force.

        Args:
            direction: [x, y, z] approach direction. Defaults to [0, 0, -1] (down).
            velocity: Approach velocity in m/s.
            max_force: Maximum contact force in N before stopping.
        """
        if direction is None:
            direction = [0.0, 0.0, -1.0]

        self._ensure_connected()
        self._pause_stream_heartbeat()
        with self._lock:
            self._switch_mode(flexivrdk.Mode.NRT_PRIMITIVE_EXECUTION)
            self._robot.ExecutePrimitive(
                "Contact",
                {
                    "contactDir": direction,
                    "contactVel": velocity,
                    "maxContactForce": max_force,
                    "enableFineContact": 1,
                },
            )

        while True:
            with self._lock:
                states = self._robot.primitive_states()
            if states.get("terminated"):
                break
            time.sleep(0.1)
        logger.info("Contact detected (max_force=%.1fN)", max_force)

    def float(
        self,
        axes: list[int] = None,
        damping: list[float] = None,
    ):
        """Enter zero-gravity floating mode. Blocking (runs until stopped).

        Args:
            axes: 6-element list, 1=floating allowed, 0=locked. Default all floating.
            damping: 6-element damping level [0-100]. Default all 0 (easiest to move).
        """
        if axes is None:
            axes = [1, 1, 1, 1, 1, 1]
        if damping is None:
            damping = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

        self._ensure_connected()
        self._pause_stream_heartbeat()
        with self._lock:
            self._switch_mode(flexivrdk.Mode.NRT_PRIMITIVE_EXECUTION)
            self._robot.ExecutePrimitive(
                "Float",
                {
                    "floatingAxis": axes,
                    "dampingLevel": damping,
                },
            )
        logger.info("Floating mode active (axes=%s, damping=%s)", axes, damping)

    # ──────────────────────────────────────────────────────────────────
    # Tool management
    # ──────────────────────────────────────────────────────────────────

    def set_tool(
        self,
        name: str,
        mass: float,
        com: list[float],
        inertia: list[float],
        tcp_location: list[float] = None,
    ):
        """Register and activate a tool for gravity compensation.

        Must be called before entering modes that need tool compensation
        (e.g. float). Robot switches to IDLE internally, then back.

        Args:
            name: Tool name (e.g. "CoinFT").
            mass: Tool mass in kg.
            com: Center of mass [x, y, z] in metres.
            inertia: Inertia [Ixx, Iyy, Izz, Ixy, Ixz, Iyz].
            tcp_location: TCP offset [x, y, z, qw, qx, qy, qz]. None = no offset.
        """
        self._ensure_connected()
        self._pause_stream_heartbeat()
        with self._lock:
            self._switch_mode(flexivrdk.Mode.IDLE)

            params = flexivrdk.ToolParams()
            params.mass = mass
            params.CoM = com
            params.inertia = inertia
            if tcp_location is not None:
                params.tcp_location = tcp_location

            if self._tool.exist(name):
                self._tool.Switch("Flange")
                self._tool.Remove(name)

            self._tool.Add(name, params)
            self._tool.Switch(name)
            logger.info("Tool active: %s (mass=%.2fkg)", name, mass)

    # ──────────────────────────────────────────────────────────────────
    # Streaming control (call at control rate)
    # ──────────────────────────────────────────────────────────────────

    def stream_cartesian(
        self,
        pose: Pose,
        wrench: list[float] = None,
        max_linear_vel: float = None,
        max_angular_vel: float = None,
        max_linear_acc: float = None,
        max_angular_acc: float = None,
    ):
        """Send a non-realtime Cartesian motion+force target.

        Call this at the control rate (e.g. 50Hz). The robot holds
        the last target if no new command arrives.

        Args:
            pose: Target pose.
            wrench: Target wrench [fx,fy,fz,tx,ty,tz]. Default all zeros.
            max_linear_vel: Override config max linear velocity (m/s).
            max_angular_vel: Override config max angular velocity (rad/s).
            max_linear_acc: Override config max linear acceleration (m/s²).
            max_angular_acc: Override config max angular acceleration (rad/s²).
        """
        self._ensure_connected()
        self._check_pose_safe(pose)
        if wrench is None:
            wrench = [0.0] * 6

        mlv = max_linear_vel or self._config.max_linear_vel
        mav = max_angular_vel or self._config.max_angular_vel
        mla = max_linear_acc or self._config.max_linear_acc
        maa = max_angular_acc or self._config.max_angular_acc

        with self._lock:
            self._switch_mode(flexivrdk.Mode.NRT_CARTESIAN_MOTION_FORCE)
            self._ensure_cartesian_contact_wrench_locked()
            self._robot.SendCartesianMotionForce(
                pose.to_rdk(), wrench, mlv, mav, mla, maa
            )
        self._note_stream_command()

    def set_impedance(
        self,
        stiffness: list[float],
        damping: list[float],
    ):
        """Set Cartesian impedance parameters.

        Args:
            stiffness: [Kx, Ky, Kz, Krx, Kry, Krz] in N/m and Nm/rad.
            damping: [Zx, Zy, Zz, Zrx, Zry, Zrz] ratio [0.3-0.8].
        """
        self._ensure_connected()
        with self._lock:
            self._robot.SetCartesianImpedance(stiffness, damping)
        logger.info("Impedance set: K=%s, Z=%s", stiffness, damping)

    def set_max_contact_wrench(self, wrench: list[float]):
        """Set maximum contact wrench before robot auto-stops.

        Args:
            wrench: [fx, fy, fz, tx, ty, tz] in N and Nm.
        """
        self._ensure_connected()
        with self._lock:
            self._robot.SetMaxContactWrench(wrench)
            self._configured_max_contact_wrench = tuple(float(value) for value in wrench)

    def set_force_control_axis(self, axes: list[bool]):
        """Set which axes use force control vs motion control.

        Args:
            axes: 6-element list. True=force control, False=motion control.
        """
        self._ensure_connected()
        with self._lock:
            self._robot.SetForceControlAxis(axes)

    # ──────────────────────────────────────────────────────────────────
    # State
    # ──────────────────────────────────────────────────────────────────

    def get_state(self) -> RobotState:
        """Read current robot state (pose, wrench, joints)."""
        self._ensure_connected()
        with self._lock:
            states = self._robot.states()
            operational = self._robot.operational()
            fault = self._robot.fault()
        return RobotState(
            pose=Pose.from_rdk(states.tcp_pose),
            wrench=Wrench.from_list(states.ext_wrench_in_tcp),
            joint_positions=list(states.q),
            timestamp=time.monotonic(),
            operational=operational,
            fault=fault,
        )

    def fault(self) -> bool:
        """True if robot is in fault state."""
        with self._lock:
            if self._robot is None:
                return False
            return self._robot.fault()

    def healthy(self) -> bool:
        """True if connected, operational, no fault, not stopped."""
        if not self._connected or self._robot is None:
            return False
        with self._lock:
            return (
                self._robot.operational()
                and not self._robot.fault()
                and not self._stopped
            )

    def stop(self):
        """Emergency stop. Robot holds position."""
        self._pause_stream_heartbeat()
        with self._lock:
            if self._robot is not None:
                self._robot.Stop()
            self._stopped = True
        logger.warning("Robot stopped")

    # ──────────────────────────────────────────────────────────────────
    # Internal helpers
    # ──────────────────────────────────────────────────────────────────

    def _ensure_connected(self):
        """Raise if not connected."""
        if not self._connected or self._robot is None:
            raise ConnectionLost("Not connected. Call connect() first.")
        if self._stopped:
            raise RobotFault("Robot is stopped. Call connect() to re-initialize.")

    def _check_pose_safe(self, pose: Pose):
        """Validate a commanded pose against synchronous safety checks."""
        try:
            self._safety.check_workspace(pose)
        except SafetyViolation:
            self.stop()
            raise

    def _start_watchdog(self):
        """Start the robot-side safety watchdog."""
        self._stop_watchdog()
        self._watchdog = SafetyWatchdog(
            config=self._config.safety,
            get_state=self._get_state_for_watchdog,
            has_fault=self._fault_for_watchdog,
            stop_robot=self.stop,
        )
        self._watchdog.start()

    def _stop_watchdog(self):
        """Stop the watchdog thread if it is running."""
        if self._watchdog is not None:
            self._watchdog.shutdown()
            self._watchdog = None

    def _pause_stream_heartbeat(self):
        """Pause heartbeat monitoring during non-streaming robot operations."""
        if self._watchdog is not None:
            self._watchdog.pause_heartbeat()

    def _note_stream_command(self):
        """Refresh streaming heartbeat after a successful Cartesian command."""
        if self._watchdog is not None:
            self._watchdog.note_stream_command()

    def _get_state_for_watchdog(self) -> Optional[RobotState]:
        """Internal state read for watchdog use.

        Returns None when the robot is disconnected or manually stopped so the
        watchdog can stay quiet during teardown and stop conditions.
        """
        with self._lock:
            if not self._connected or self._robot is None or self._stopped:
                return None
            states = self._robot.states()
            operational = self._robot.operational()
            fault = self._robot.fault()
        return RobotState(
            pose=Pose.from_rdk(states.tcp_pose),
            wrench=Wrench.from_list(states.ext_wrench_in_tcp),
            joint_positions=list(states.q),
            timestamp=time.monotonic(),
            operational=operational,
            fault=fault,
        )

    def _fault_for_watchdog(self) -> bool:
        """Internal robot fault read for watchdog use."""
        with self._lock:
            if not self._connected or self._robot is None or self._stopped:
                return False
            return self._robot.fault()

    def _switch_mode(self, target_mode: flexivrdk.Mode):
        """Switch RDK mode if not already in it. Must hold self._lock."""
        if self._mode == target_mode:
            return
        self._robot.SwitchMode(target_mode)
        self._mode = target_mode
        if target_mode != flexivrdk.Mode.NRT_CARTESIAN_MOTION_FORCE:
            self._configured_max_contact_wrench = None

    def _ensure_cartesian_contact_wrench_locked(self):
        """Apply configured contact-wrench regulation in Cartesian streaming mode."""
        target_wrench = tuple(
            float(value) for value in self._config.safety.max_contact_wrench
        )
        if self._configured_max_contact_wrench == target_wrench:
            return
        self._robot.SetMaxContactWrench(list(target_wrench))
        self._configured_max_contact_wrench = target_wrench

File: ros2_ws/src/arm_commander/src/arm_commander.cpp
/**
 * @file arm_commander.cpp
 * @brief ArmCommander implementation — single RDK owner.
 */

#include "arm_commander/arm_commander.hpp"

#include <Eigen/Geometry>
#include <spdlog/spdlog.h>

#include <chrono>
#include <cmath>
#include <stdexcept>
#include <thread>

namespace arm_commander {

ArmCommander::ArmCommander(const RobotConfig& config) : config_(config) {}

ArmCommander::~ArmCommander()
{
    try {
        shutdown();
    } catch (const std::exception& e) {
        spdlog::error("Exception in ~ArmCommander: {}", e.what());
    }
}

// ── Lifecycle ──────────────────────────────────────────────────────────────

void ArmCommander::connect()
{
    std::lock_guard<std::mutex> lock(mutex_);

    robot_ = std::make_unique<flexiv::rdk::Robot>(config_.serial_number);

    if (robot_->fault()) {
        spdlog::warn("Fault detected — clearing...");
        if (!robot_->ClearFault()) {
            throw std::runtime_error("Cannot clear robot fault");
        }
        std::this_thread::sleep_for(std::chrono::seconds(2));
        if (robot_->fault()) {
            throw std::runtime_error("Cannot clear fault — check E-stop and teach pendant");
        }
        spdlog::info("Fault cleared");
    }

    spdlog::info("Enabling robot...");
    robot_->Enable();

    auto start = std::chrono::steady_clock::now();
    while (!robot_->operational()) {
        if (std::chrono::steady_clock::now() - start > std::chrono::seconds(10)) {
            throw std::runtime_error("Timeout waiting for robot to become operational");
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    tool_ = std::make_unique<flexiv::rdk::Tool>(*robot_);
    current_mode_ = flexiv::rdk::Mode::UNKNOWN;
    connected_ = true;
    stopped_ = false;
    stop_requested_ = false;
    spdlog::info("Robot operational: {}", config_.serial_number);
}

void ArmCommander::shutdown()
{
    std::lock_guard<std::mutex> lock(mutex_);
    if (robot_) {
        try {
            robot_->Stop();
        } catch (const std::exception& e) {
            spdlog::warn("Error during shutdown Stop(): {}", e.what());
        }
    }
    connected_ = false;
    stopped_ = true;
    spdlog::info("Robot shutdown complete");
}

// ── Blocking primitives ────────────────────────────────────────────────────

void ArmCommander::home()
{
    ensure_connected();
    {
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::NRT_PLAN_EXECUTION);
        robot_->ExecutePlan("PLAN-Home");
    }
    wait_busy();
    spdlog::info("Home position reached");
}

void ArmCommander::zero_ft()
{
    ensure_connected();
    {
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::NRT_PRIMITIVE_EXECUTION);
        robot_->ExecutePrimitive("ZeroFTSensor",
            std::map<std::string, flexiv::rdk::FlexivDataTypes>());
    }
    wait_primitive("terminated");
    spdlog::info("FT sensor zeroed");
}

void ArmCommander::contact(
    const std::vector<double>& direction, double velocity, double max_force)
{
    ensure_connected();
    {
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::NRT_PRIMITIVE_EXECUTION);
        std::map<std::string, flexiv::rdk::FlexivDataTypes> params;
        params["contactDir"] = std::vector<double>(direction.begin(), direction.end());
        params["contactVel"] = velocity;
        params["maxContactForce"] = max_force;
        params["enableFineContact"] = 1;
        robot_->ExecutePrimitive("Contact", params);
    }
    wait_primitive("terminated");
    spdlog::info("Contact detected (max_force={:.1f}N)", max_force);
}

void ArmCommander::move_to(const std::array<double, 7>& pose, double velocity)
{
    ensure_connected();
    check_workspace(pose[0], pose[1], pose[2]);

    // Convert quaternion [qw,qx,qy,qz] to Euler ZYX degrees for MoveL Coord
    Eigen::Quaterniond q(pose[3], pose[4], pose[5], pose[6]);  // w,x,y,z
    Eigen::Vector3d euler = q.toRotationMatrix().eulerAngles(2, 1, 0);  // ZYX
    double rx_deg = euler[2] * 180.0 / M_PI;
    double ry_deg = euler[1] * 180.0 / M_PI;
    double rz_deg = euler[0] * 180.0 / M_PI;

    {
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::NRT_PRIMITIVE_EXECUTION);

        std::map<std::string, flexiv::rdk::FlexivDataTypes> params;
        params["target"] = flexiv::rdk::Coord(
            {pose[0], pose[1], pose[2]}, {rx_deg, ry_deg, rz_deg},
            {"WORLD", "WORLD_ORIGIN"});
        params["vel"] = velocity;
        robot_->ExecutePrimitive("MoveL", params);
    }
    wait_primitive("reachedTarget");
    spdlog::info("MoveL complete");
}

// ── Tool management ────────────────────────────────────────────────────────

// TODO: register CoinFT as a tool (mass / CoM / inertia in robot.yaml) so gravity
// compensation works in floating mode and Cartesian impedance control.
void ArmCommander::set_tool(const std::string& name, double mass,
    const std::array<double, 3>& com, const std::array<double, 6>& inertia,
    const std::array<double, 7>& tcp_location)
{
    ensure_connected();
    std::lock_guard<std::mutex> lock(mutex_);
    switch_mode(flexiv::rdk::Mode::IDLE);

    flexiv::rdk::ToolParams params;
    params.mass = mass;
    params.CoM = com;
    params.inertia = inertia;
    if (tcp_location != std::array<double, 7>{}) {
        params.tcp_location = tcp_location;
    }

    if (tool_->exist(name)) {
        tool_->Switch("Flange");
        tool_->Remove(name);
    }

    tool_->Add(name, params);
    tool_->Switch(name);
    spdlog::info("Tool active: {} (mass={:.2f}kg)", name, mass);
}

// ── Streaming control ──────────────────────────────────────────────────────

void ArmCommander::stream_cartesian(
    const std::array<double, 7>& pose, const std::array<double, 6>& wrench,
    double max_linear_vel, double max_angular_vel,
    double max_linear_acc, double max_angular_acc)
{
    ensure_connected();
    check_workspace(pose[0], pose[1], pose[2]);

    // Use caller overrides if provided (> 0), else fall back to config
    double lin_vel = max_linear_vel > 0.0 ? max_linear_vel : config_.max_linear_vel;
    double ang_vel = max_angular_vel > 0.0 ? max_angular_vel : config_.max_angular_vel;
    double lin_acc = max_linear_acc > 0.0 ? max_linear_acc : config_.max_linear_acc;
    double ang_acc = max_angular_acc > 0.0 ? max_angular_acc : config_.max_angular_acc;

    std::lock_guard<std::mutex> lock(mutex_);
    switch_mode(flexiv::rdk::Mode::NRT_CARTESIAN_MOTION_FORCE);
    robot_->SendCartesianMotionForce(pose, wrench, lin_vel, ang_vel, lin_acc, ang_acc);
}

void ArmCommander::set_impedance(
    const std::array<double, 6>& stiffness, const std::array<double, 6>& damping)
{
    ensure_connected();
    std::lock_guard<std::mutex> lock(mutex_);
    robot_->SetCartesianImpedance(stiffness, damping);
    spdlog::info("Impedance set");
}

void ArmCommander::set_max_contact_wrench(const std::array<double, 6>& wrench)
{
    ensure_connected();
    std::lock_guard<std::mutex> lock(mutex_);
    robot_->SetMaxContactWrench(wrench);
}

void ArmCommander::set_force_control_axis(const std::array<bool, 6>& axes)
{
    ensure_connected();
    std::lock_guard<std::mutex> lock(mutex_);
    robot_->SetForceControlAxis(axes);
}

void ArmCommander::set_force_control_frame(flexiv::rdk::CoordType frame)
{
    ensure_connected();
    std::lock_guard<std::mutex> lock(mutex_);
    robot_->SetForceControlFrame(frame);
}

// ── RT joint torque floating ───────────────────────────────────────────────

void ArmCommander::float_joints(const std::vector<double>& damping_gains)
{
    ensure_connected();

    // Capture robot pointer for the periodic task (no lock needed inside
    // scheduler — RDK streaming calls are designed for single-thread RT use)
    auto* robot_ptr = robot_.get();
    const auto& gains = damping_gains;
    std::atomic<bool> sched_stop{false};

    auto periodic_task = [robot_ptr, &gains, &sched_stop, this]() {
        try {
            if (robot_ptr->fault()) {
                spdlog::error("Fault during floating");
                sched_stop = true;
                return;
            }

            std::vector<double> torque(robot_ptr->info().DoF, 0.0);
            for (size_t i = 0; i < torque.size() && i < gains.size(); ++i) {
                torque[i] = -gains[i] * robot_ptr->states().dtheta[i];
            }

            robot_ptr->StreamJointTorque(torque, true, true);
        } catch (const std::exception& e) {
            spdlog::error("Float periodic: {}", e.what());
            sched_stop = true;
        }
    };

    // Set up scheduler before switching mode to minimize gap between
    // entering RT_JOINT_TORQUE and first torque command
    flexiv::rdk::Scheduler scheduler;
    scheduler.AddTask(periodic_task, "floating", 1, scheduler.max_priority());

    {
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::RT_JOINT_TORQUE);
    }

    scheduler.Start();
    spdlog::info("Floating mode active — move the robot by hand");

    // Block until stop requested
    while (!sched_stop && !stop_requested_) {
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    scheduler.Stop();
    spdlog::info("Floating stopped");
}

// ── Cartesian floating ─────────────────────────────────────────────────────

void ArmCommander::float_cartesian(double loop_rate_hz)
{
    (void)loop_rate_hz;  // RT scheduler runs at 1kHz internally
    ensure_connected();

    // Send one NRT command to enter Cartesian mode so we can configure impedance
    {
        auto state = get_state();
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::NRT_CARTESIAN_MOTION_FORCE);
        robot_->SendCartesianMotionForce(state.tcp_pose, {0, 0, 0, 0, 0, 0});
    }

    // Zero stiffness = free-floating on all Cartesian axes
    set_force_control_axis({false, false, false, false, false, false});
    set_impedance({0, 0, 0, 0, 0, 0}, {0.3, 0.3, 0.3, 0.3, 0.3, 0.3});

    // Now switch to RT mode for true free-floating (NRT motion generator fights pushes)
    auto* robot_ptr = robot_.get();
    std::atomic<bool> sched_stop{false};

    auto periodic_task = [robot_ptr, &sched_stop, this]() {
        try {
            if (robot_ptr->fault()) {
                spdlog::error("Fault during Cartesian floating");
                sched_stop = true;
                return;
            }

            // Stream current pose back as target — with K=0 the position
            // target is irrelevant but keeps the RT stream alive
            auto tcp = robot_ptr->states().tcp_pose;
            robot_ptr->StreamCartesianMotionForce(tcp, {0, 0, 0, 0, 0, 0});
        } catch (const std::exception& e) {
            spdlog::error("Cartesian float periodic: {}", e.what());
            sched_stop = true;
        }
    };

    flexiv::rdk::Scheduler scheduler;
    scheduler.AddTask(periodic_task, "cartesian_float", 1, scheduler.max_priority());

    {
        std::lock_guard<std::mutex> lock(mutex_);
        switch_mode(flexiv::rdk::Mode::RT_CARTESIAN_MOTION_FORCE);
    }

    scheduler.Start();
    spdlog::info("Cartesian floating active (RT) — move the robot by hand");

    while (!sched_stop && !stop_requested_) {
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    scheduler.Stop();
    spdlog::info("Cartesian floating stopped");
}

// ── State ──────────────────────────────────────────────────────────────────

RobotState ArmCommander::get_state()
{
    ensure_connected();
    std::lock_guard<std::mutex> lock(mutex_);
    RobotState state;
    auto s = robot_->states();
    for (size_t i = 0; i < 7 && i < s.tcp_pose.size(); ++i) {
        state.tcp_pose[i] = s.tcp_pose[i];
    }
    for (size_t i = 0; i < 6 && i < s.ext_wrench_in_tcp.size(); ++i) {
        state.wrench[i] = s.ext_wrench_in_tcp[i];
    }
    for (size_t i = 0; i < 6 && i < s.ext_wrench_in_world.size(); ++i) {
        state.wrench_in_world[i] = s.ext_wrench_in_world[i];
    }
    state.joint_positions = std::vector<double>(s.q.begin(), s.q.end());
    state.joint_velocities = std::vector<double>(s.dtheta.begin(), s.dtheta.end());
    state.operational = robot_->operational();
    state.fault = robot_->fault();
    return state;
}

bool ArmCommander::has_fault()
{
    std::lock_guard<std::mutex> lock(mutex_);
    if (!robot_)
        return false;
    return robot_->fault();
}

bool ArmCommander::is_healthy()
{
    if (!connected_ || !robot_)
        return false;
    std::lock_guard<std::mutex> lock(mutex_);
    return robot_->operational() && !robot_->fault() && !stopped_;
}

void ArmCommander::stop()
{
    std::lock_guard<std::mutex> lock(mutex_);
    if (robot_) {
        robot_->Stop();
    }
    stopped_ = true;
    stop_requested_ = true;
    spdlog::warn("Robot stopped");
}

// ── Internal ───────────────────────────────────────────────────────────────

void ArmCommander::ensure_connected()
{
    if (!connected_ || !robot_) {
        throw std::runtime_error("Not connected. Call connect() first.");
    }
    if (stopped_) {
        throw std::runtime_error("Robot stopped. Call connect() to re-initialize.");
    }
}

void ArmCommander::check_workspace(double x, double y, double z)
{
    if (!config_.safety.workspace.contains(x, y, z)) {
        stop();
        throw std::runtime_error("Pose outside workspace bounds");
    }
}

void ArmCommander::switch_mode(flexiv::rdk::Mode target)
{
    // Must be called with mutex_ held
    if (current_mode_ == target)
        return;
    robot_->SwitchMode(target);
    current_mode_ = target;
}

void ArmCommander::wait_busy()
{
    while (!stop_requested_) {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            if (!robot_->busy())
                return;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    throw std::runtime_error("Interrupted by stop request");
}

void ArmCommander::wait_primitive(const std::string& state_key)
{
    // Most primitives don't exit by themselves — busy() stays true.
    // Must poll primitive_states() for the expected state key.
    while (!stop_requested_) {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            auto states = robot_->primitive_states();
            auto it = states.find(state_key);
            if (it != states.end()) {
                if (auto* val = std::get_if<int>(&it->second)) {
                    if (*val)
                        return;
                }
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    throw std::runtime_error("Interrupted by stop request");
}

}  // namespace arm_commander
